<a href="https://colab.research.google.com/github/lapshinaaa/recsys-tasks/blob/main/RecSys3_CanGen.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Recommender Systems
## Candidate Generation

Note from YSDA course: During practice, we've discussed common CG stage abstractions, quality metrics and implemented several models with a KNN candidate generator class. In homework your task will be to dive deeper into this domain and implement a few more models, compare them and come up with a way to use all of them simultaneously. The homework consists of several subtasks, which are independent, and two bonus subtasks.

Let's go!

### Imports and metrics setup
We'll use the metrics and evaluator class, which we were using in practice about candidate generators.

In [1]:
# if you are running this notebook in colab, datasphere, etc - you need to clone the repository first

!git clone https://github.com/yandexdataschool/recsys_course
!pip install -r recsys_course/week03/homework/requirements.txt
!pip install -e recsys_course/grocery

Cloning into 'recsys_course'...
remote: Enumerating objects: 206, done.
remote: Counting objects: 100% (17/17), done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 206 (delta 8), reused 5 (delta 4), pack-reused 189 (from 1)
Receiving objects: 100% (206/206), 34.65 MiB | 23.65 MiB/s, done.
Resolving deltas: 100% (75/75), done.
ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'recsys_course/week03/homework/requirements.txt'
Obtaining file:///content/recsys_course/grocery
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Installing backend dependencies ... done
  Preparing editable metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 2.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.2/99.2 MB 9.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 117.2 MB/s eta 0

In [1]:
import numba as nb
import numpy as np
import polars as pl
import scipy.sparse as sp
from sklearn.model_selection import train_test_split

from grocery.metrics import Evaluator, Recall, Novelty, CategoryDiversity
from grocery.recommender.candidates import CandidateGenerator, Candidate
from grocery.utils.dataset import build_matrix_with_mappings
from collections import defaultdict
from scipy.sparse.linalg import inv
from grocery.utils.dataset import download_and_extract

In [2]:
DATA_DIR = "../../data/lavka"

download_and_extract(
     url="https://www.kaggle.com/api/v1/datasets/download/thekabeton/ysda-recsys-2025-lavka-dataset",
     filename="lavka.zip",
     dest_dir=DATA_DIR
 )

lavka.zip: 100%|██████████████████████████████████████████████████| 447M/447M [00:12<00:00, 36.4MB/s]


Unpacking lavka.zip...
Files from lavka.zip successfully unpacked



In [3]:
data = (
    pl.read_parquet(f"{DATA_DIR}/train.parquet")
    .filter(pl.col("action_type") == "AT_CartUpdate")
    .select(
        pl.col("user_id"),
        pl.col("product_id").alias("item_id"),
        pl.col("request_id"),
        pl.lit(1).alias("rating"),
        pl.col("timestamp"),
        pl.col("product_category"),
    )
)

# global timepoint split!
train, test = train_test_split(data.sort("timestamp"), test_size=0.2, shuffle=False)

In [4]:
evaluator = Evaluator(
    metrics=[
        Recall(10),
        Recall(50),
        Novelty(train, 10),
        Novelty(train, 50),
        CategoryDiversity(train, 10),
        CategoryDiversity(train, 50),
    ]
)

evaluator.load_test_actions(
    test.filter(pl.col("user_id").is_in(train["user_id"].unique()))
)

/tmp/ipython-input-4038882061.py:13: DeprecationWarning: `is_in` with a collection of the same datatype is ambiguous and deprecated.
Please use `implode` to return to previous behavior.

See https://github.com/pola-rs/polars/issues/22149 for more information.
  test.filter(pl.col("user_id").is_in(train["user_id"].unique()))


## Candidate Generators

### EASE (0.5 points), LinearItemToItemCG and ANN (1 points)
Implement the model from [Embarrassingly Shallow Autoencoders for Sparse Data](https://arxiv.org/pdf/1905.03375).

From practice we know that we can present our training data as a sparse matrix $X \in \mathbb{R}^{|U| x |I|}$, where each element $x_{ui}$ is a binary value, which is positive if the user $u$ had an interaction with item $i$, or rating from their interactions, or number of interactions.

EASE suggests a linear model, that encodes all items as an item-item matrix $B$, and each item can be represented as a linear combination of other items.
We consider two features. The first is categorical — the ID of the item being evaluated. The second group consists of items with which the user has had positive interactions (excluding the item in question). We then take the cross-product of these two groups to generate features like: [currently evaluating item i, item j appears in the user’s history]. We train a linear model on these cross-features. The target is 1 for positive interactions and 0 otherwise (even if the user hasn’t been shown the item).

We use MSE loss and L2 regularization (a convex loss function), adding the constraint that diagonal weights must be zero. This constraint, $diag(B) = 0$, is crucial as to avoid the trivial solution $B = I$ (self-similarity of items), where $I$ is the identity matrix.

The optimization problem in matrix form is as follows:

$$\min_{B} ||X - XB||^2_F + \lambda ||B||^2_F$$
$$\text{s.t. diag}(B) = 0$$

Turns out, the problem has an exact analytical solution - your task is to derive it and implement!

*Notes: it's better to implement both sparse with scipy.sparse.linalg.inv approach and dense with np.linalg.inv, and to check if dense matrix fits in your memory :-)*

In [5]:
class EASE:
    def __init__(self, lambd: float = 0.1):
        self.lambd = lambd
        self.B = None
        self.item_id2idx = None # column index -> item_id
        self.item_idx2id = None # item_id -> column index
        self.user_id2idx = None
        self.user_idx2id = None
        self.X = None

    def fit(self, ratings: pl.DataFrame, sparse_inv: bool = False):
        X, (user_id2idx, item_id2idx, user_idx2id, item_idx2id) = build_matrix_with_mappings(ratings)
        self.X = X.tocsr() # sparsify
        self.item_id2idx = item_id2idx
        self.item_idx2id = item_idx2id
        self.user_id2idx = user_id2idx
        self.user_idx2id = user_idx2id

        # Gram Matrix G (according to the article) = X^T X
        G = X.T @ X  # shape: (num_items x num_items), sparse

        # λ * I for regularization
        num_items = G.shape[0]
        G = G + self.lambd * sp.eye(num_items)
        G = G.tocsc() # converting to csc format for optimal calculations

        P = inv(G)  # returns sparse linear operator
        P = P.toarray()  # convert to dense (safe because only num_items x num_items)

        #  B
        B = -P / np.diag(P)  # broadcasting division by diagonal
        np.fill_diagonal(B, 0)  # enforce diag(B)=0

        self.B = B
        return self

Now let's implement a class to use that model for extracting candidates (or straight-up recommending) for a certain user.

Some notes:
- Do not forget about the fact that model operates in indices, but CG should receive user IDs and return `Candidates` with item IDs (not indices!)
- Utilize the fact that user-item interactions is a sparse matrix, meaning there are only few positive elements, so you can try to iterate over them, extract needed weights from sparse matrix and sum them.
- You may not even store the user-item matrix, but receive a dict with user history as a list of IDs during `__init__.py`. Although this is not a part of this homework, such class would later allow you to switch the dict for some other user history storage (or maybe even a generator, yeilding results from a database, .etc).

In [5]:
class LinearItemToItemCG(CandidateGenerator):
    def __init__(self,
                 B: sp.spmatrix | np.ndarray,
                 user_id2idx: dict[int, int],
                 item_id2idx: dict[int, int],
                 X: sp.spmatrix | np.ndarray,
                #  history: dict[int, list[int]],
                 ):
        """Item to Item Linear Model applier.

        Args:
            B (sp.sparse.spmatrix | np.ndarray): Weight matrix.
            user_id2idx (dict[int, int]): Mapping from user IDs to indices.
            item_id2idx (dict[int, int]): Mapping from item IDs to indices.
            X (sp.sparse.spmatrix | np.ndarray): User-item interaction matrix.
        """
        super().__init__()

        self.B = B                          # (num_items × num_items)
        self.user_id2idx = user_id2idx      # map user_id → row index in X
        self.item_id2idx = item_id2idx      # map item_id → column index in X
        self.idx2item_id = {v: k for k, v in item_id2idx.items()} # reversing the mapping here
        self.X = X.tocsr()


    def extract_candidates(self, object_id: int, n: int = 10) -> list[Candidate]:

        if object_id not in self.user_id2idx: # edge case user does not exist
          return []

        user_idx = self.user_id2idx[object_id]
        user_vector = self.X[user_idx].tocsr() # getting the user history as a sparse row
        B = self.B.tocsr()

        # now we're going to score the items and remove the ones the user has already interacted with
        scores = (user_vector @ B).toarray().ravel()
        seen_items = set(user_vector.indices)
       # print(scores) debuggins that fried my brain
        scores[list(seen_items)] = -np.inf # we're not recommending what has already been seen

        # select top items
        top_indices = np.argpartition(-scores, n)[:n]
        top_indices = top_indices[np.argsort(-scores[top_indices])]

        return [Candidate(id=self.idx2item_id[idx]) for idx in top_indices]

Your model should meet the quality level needed - for defined in this notebook train dataset, initialization and default hyperparameters from model class template, reference implementation achieves:

$$recall@10 >= 0.04$$
$$recall@50 >= 0.15$$

In [ ]:
model = EASE(lambd=0.1)
model.fit(train)

In [ ]:
ease_cg = LinearItemToItemCG(model.B, model.user_id2idx, model.item_id2idx, model.X)

metrics = evaluator.evaluate(lambda user_id, n: ease_cg.extract_candidates(user_id, n))
for metric_name, metric_value in metrics.items():
    print(f"{metric_name} = {metric_value:.3f}")

Now let's implement ANN candgen and compare searching in ALS embedding space with EASE scoring. There are lots of options to choose from, but we recommend settling on `faiss` or `voyager` libraries. The latter utilizes HNSW algorithm and has `ef_construction` parameter when building and `query_ef` parameter in `query` method, they control the tradeoff between speed and recall - feel free to experiment a bit with them.

Here I could say *"you can write your own HNSW"*, but this is breaking a butterfly upon the wheel - please don't. The fact that ANN is used instead of KNN is enough optimization :D

In [6]:
from grocery.models import ALS
from grocery.recommender.candidates import DotProductKNN

def extract_embeddings(model: ALS) -> tuple[dict[int, np.array], dict[int, np.array]]:
    return (
        {ID: model.user_vectors[idx] for ID, idx in model.user_id2idx.items()},
        {ID: model.item_vectors[idx] for ID, idx in model.item_id2idx.items()}
    )

In [26]:
from voyager import Index, Space
# what space do you need to use? - cosine


class ANN(CandidateGenerator):
    def __init__(self,
                 left_embeddings: dict[int, np.array],
                 right_embeddings: dict[int, np.array]):
        super().__init__()

        self.item_ids = list(right_embeddings.keys())
        self.item_vectors = np.vstack([right_embeddings[i] for i in self.item_ids]) # (num_items x embedding_dim)

        # build HNSW index (Hierarchical Navigable Small World Graph)
        self.index = self.build_index(self.item_vectors, self.item_ids)


    @staticmethod
    def build_index(vectors: np.ndarray, ids: list[int]):
        dim=vectors.shape[1]
        capacity=len(ids)

        index = Index(space=Space.Cosine, num_dimensions=dim, max_elements=capacity)
        index.add_items(vectors, ids)

        return index

    def extract_candidates(self, object_id: int, n: int = 10) -> list[Candidate]:
        query = self.item_vectors[self.item_ids.index(object_id)]
        # first, we're going to find nearest neighbors for this candidate
        ids, dists = self.index.query(query, k=n+1)

        # remove the item itself because distance can't be equal to 0
        ids =[cand_id for cand_id in ids if cand_id != object_id]

        return [Candidate(id=i) for i in ids[:n]]

Compare the resulting class with DotProductKNN from practice: speed, recall.

In [27]:
als = ALS(max_iter=100, dim=64, lr=0.01, reg_embeddings=1e-4)
als.fit(train)

100%|██████████| 100/100 [00:17<00:00,  5.71it/s]


In [28]:
user_embeddings, item_embeddings = extract_embeddings(als)
als_ann = ANN(user_embeddings, item_embeddings)
als_knn = DotProductKNN(user_embeddings, item_embeddings)

In [ ]:
user_history = (  # here I'm buildinh user -> items history
    train
    .group_by("user_id")
    .agg(pl.col("item_id").unique().alias("items"))
    .to_dict(as_series=False)
)
user2items = {uid: set(items) for uid, items in zip(user_history["user_id"], user_history["items"])}

In [29]:
# here I am adapting ANN since it's item -> item:  make ANN (item→item) behave like user→item for the evaluator
def ann_user_adapter(user_id: int, n: int = 10):
    seen = user2items.get(user_id, set())
    if not seen:
        return []

    # how many neighbors to pull per seed item
    K_PER_ITEM = max(n * 5, 50)

    scores = defaultdict(int)

    # vote-based fusion of neighbors from each seen item
    for item_id in seen:
        cands = als_ann.extract_candidates(item_id, n=K_PER_ITEM)
        for c in cands:
            if c.id not in seen:
                scores[c.id] += 1

    if not scores:
        return []

    # rank by votes (desc), tie-break by item id for determinism
    ranked = sorted(scores.items(), key=lambda kv: (-kv[1], kv[0]))
    top_ids = [item_id for item_id, _ in ranked[:n]]

    return [Candidate(id=i) for i in top_ids]


metrics_als_ann = evaluator.evaluate(lambda user_id, k: ann_user_adapter(user_id, k))
for metric_name, metric_value in metrics_als_ann.items():
    print(f"{metric_name} = {metric_value:.3f}")

100%|██████████| 31763/31763 [2:35:04<00:00,  3.41it/s]


recall@10 = 0.008
recall@50 = 0.025
novelty@10 = 0.970
novelty@50 = 0.978
category_diversity@10 = 0.914
category_diversity@50 = 0.835


This item2item approach can be found in the original Amazon 2003 RecSys paper:

> We base recommendations on items-to-item collaborative filtering. At request time, user histories are expanded via item-item similarity.



### 🔍 ANN Candidate Generation — Why We Use a User Adapter

Our ANN model retrieves similar items given an item:
ANN.extract_candidates(item_id) → list of similar item_ids

But the evaluator expects a user → item recommender:
recommender(user_id) → list of recommended item_ids

To connect these, we build a user-history adapter:

1. Take items the user interacted with in TRAIN.
2. For each of those items, get ANN nearest neighbors.
3. Merge those neighbors (vote-based aggregation).
4. Remove items the user has already seen.
5. Recommend the top-scoring items.

This turns item→item similarity into user→item recommendations, without retraining ANN.

---

#### Why This Makes Sense in Real Systems

This mirrors how large-scale production recommenders work:

| Stage | Input | Output | Purpose |
|------|-------|--------|---------|
| Candidate Generation (ANN) | item_id | similar item_ids | Fast, approximate retrieval |
| User Adapter (History-Based Expansion) | user_id | candidate items | Converts item→item to user→item |
| Reranker (e.g., CatBoost) | user + items | final ranked recommendations | Personalized scoring |

This two-stage design is used by Netflix, Amazon, YouTube, etc., because:

• ANN is extremely fast  
• User history captures preference structure  
• Rerankers later refine recommendations

So the adapter is not a hack — it is the *correct architectural step* to turn ANN into a user recommender.

### SLIM (1 point)

The next model in our arsenal will be **S**parse **Li**near **M**ethod a.k.a. [SLIM](https://ieeexplore.ieee.org/document/6137254) (PDF is available [here](https://www.researchgate.net/profile/George_Karypis/publication/220765374_SLIM_Sparse_Linear_Methods_for_Top-N_Recommender_Systems/links/549ee9ac0cf257a635fe7010.pdf)).

Actually, SLIM was introduced long before EASE, and EASE cites SLIM (and uses a lot of it as a foundation), not the other way around. For educational purposes, we switched their order of presentation in the homework. The algorithm solves the same problem of finding best item-item weight matrix, but with different loss - the main difference is L1 regularization for sparsity component, which helps the algorithm to build a sparse weight matrix, which is both memory-efficient and compute-efficient, if you organize your storage accordingly.


The optimization problem:

$$\min_{B} \frac{1}{2}||X - XB||^2_F + \frac{\beta}{2} ||B||^2_F + \lambda ||B||_F$$
$$\text{subject to } B \geq 0 \text{, diag}(B) = 0$$

This is solvable with coordinate descent by projecting each individual weight (non-negative constraint) and manually setting diagonal weights to zero during all the steps. Moreover, each item weight vector is independent from others, so the task is easily parallelizable across items. However, there aren't any optimizers that can do that out of the box. As proposed in this [paper](https://www.slideshare.net/slideshow/efficient-slides/27138952), we'll simplify the problem a bit by dropping the $diag(B) = 0$ constraint and setting the item column of the interaction matrix to zero when fitting for it. Let's see, whether it's still a viable model - simple scikit-learn should do the trick.

Implement SLIM model using sklearn ElasticNet solver and optionally multiprocessing. The steps are as follows:
- Build sparse matrix of interactions
- For each item $u$:
    - Get interaction matrix $X$ - the features
    - Extract target vector $y = X_u \in R^{|U|}$ - all interactions with this item
    - Find solution and save it
- Extract sparse coefs to one weight matrix B

*Notes:* multiprocessing and jupyter notebooks are not really friends, so if you want to speed up the model training by parallelization - move it into a separate file, save matrix to `.npy` and launch it from here using Ipython magic.

In [6]:
from tqdm import trange
from sklearn.linear_model import ElasticNet


def _solve_item(X, item_idx, alpha, l1_ratio):

    # TODO: your code here
    X_ = X.copy().tocsr()
    X_[:, item_idx] = 0.0 # this emulates diag(B) = 0 so that the model cannot select the self-feature

    y = X[:, item_idx].toarray().ravel() # target

    solver = ElasticNet(
        alpha=alpha,
        l1_ratio=l1_ratio,
        fit_intercept=False,
        positive=True, # enforcing that B >= 0
        max_iter=1000,
        tol=1e-4,
        selection='random' # that is a standard coord-descent option
    )

    solver.fit(X_, y)
    coef = solver.coef_.astype(np.float32)
    coef[item_idx] = 0.0 # enforcing zero on the diagonal explicitly

    return coef


class SLIM:
    def __init__(self,
                 alpha: float = 0.1,
                 l1_ratio: float = 0.01,
                 num_processes: int | None = None,
                 ):

        self.alpha = alpha
        self.l1_ratio = l1_ratio
        self.num_processes = num_processes

        self.B = None  # a learned item2item weight matrix
        self.mappings = None
        self.X = None

    def fit(self, ratings: pl.DataFrame):

        # TODO: your code here

        X, mappings = build_matrix_with_mappings(ratings)
        self.X = X.tocsr()
        self.mappings = mappings

        n_users, n_items = self.X.shape
        self.B = sp.lil_matrix((n_items, n_items), dtype=np.float32)
        for j in trange(n_items, desc='Fitting SLIM (ElasticNet)'):
          coef_j = _solve_item(self.X, j, self.alpha, self.l1_ratio)

          self.B[j, :] = coef_j

        self.B = self.B.tocsr()

        return self

In [17]:
slim = SLIM(num_processes=1)
slim.fit(train)

Streaming output truncated to the last 5000 lines.
Fitting SLIM (ElasticNet):  79%|███████▉  | 9603/12102 [17:49<04:41,  8.87it/s]/usr/local/lib/python3.12/dist-packages/scipy/sparse/_index.py:210: SparseEfficiencyWarning: Changing the sparsity structure of a csr_array is expensive. lil and dok are more efficient.
  self._set_arrayXarray(i, j, x)
Fitting SLIM (ElasticNet):  79%|███████▉  | 9604/12102 [17:49<04:50,  8.61it/s]/usr/local/lib/python3.12/dist-packages/scipy/sparse/_index.py:210: SparseEfficiencyWarning: Changing the sparsity structure of a csr_array is expensive. lil and dok are more efficient.
  self._set_arrayXarray(i, j, x)
Fitting SLIM (ElasticNet):  79%|███████▉  | 9605/12102 [17:49<05:01,  8.28it/s]/usr/local/lib/python3.12/dist-packages/scipy/sparse/_index.py:210: SparseEfficiencyWarning: Changing the sparsity structure of a csr_array is expensive. lil and dok are more efficient.
  self._set_arrayXarray(i, j, x)
Fitting SLIM (ElasticNet):  79%|███████▉  | 9606/12102 

In [64]:
user_id2idx, item_id2idx, user_idx2id, item_idx2id = slim.mappings # unpacking mappings

slim_cg = LinearItemToItemCG(B=slim.B,
                             user_id2idx=user_id2idx,
                             item_id2idx=item_id2idx,
                             X=slim.X)

In [41]:
scores = (slim.X[0] @ slim.B).toarray().ravel()
print(scores.shape)

(12102,)


In [47]:
print(slim.B.shape)
print(slim.X.shape)

(12102, 12102)
(1905, 12102)


In [65]:
metrics_slim = evaluator.evaluate(lambda user_id, n: slim_cg.extract_candidates(user_id, n))
for metric_name, metric_value in metrics_slim.items():
    print(f"{metric_name} = {metric_value:.4f}")

100%|██████████| 31763/31763 [01:04<00:00, 489.38it/s]


recall@10 = 0.0119
recall@50 = 0.0370
novelty@10 = 0.9457
novelty@50 = 0.9564
category_diversity@10 = 0.8925
category_diversity@50 = 0.8070


# SLIM Candidate Generation — Summary and Analysis

In this notebook, I implemented and evaluated the SLIM (Sparse Linear Methods) model for item-to-item recommendation. The objective of SLIM is to learn a sparse item–item similarity matrix, which can then be used to generate recommendations by finding items most similar to those a user has interacted with.

---

## ✅ What Was Implemented

1. Built the User–Item Matrix  
   I constructed a sparse interaction matrix X from the training data using binary cart-update interactions.  
   Shape example:  
   - Users: ~1900  
   - Items: ~12000  
   → Matrix X ∈ R^(1905 × 12102)

2. Implemented SLIM Using ElasticNet (Coordinate Descent)
   For each item j:
   - The target y was X[:, j] (all users who interacted with item j).
   - The feature matrix X was used as predictors, except we masked the j-th column to enforce diag(B)=0.
   - We trained ElasticNet with:
     - positive=True (B ≥ 0 constraint)
     - l1_ratio controlling sparsity
     - alpha controlling shrinkage
   - The learned coefficients formed column j of the item–item weight matrix B.

3. Built `LinearItemToItemCG` Candidate Generator
   Using:
   - B (item similarity matrix)
   - X (user interaction matrix)
   - user↔index and item↔index mappings  

And then generated releveant recs.

## 🐞 Debugging Notes

During evaluation, I encountered:
- Sparse matrix returned as coo_array, which does not support `.indices`.
- Item masking step caused dimension mismatch errors.

These were resolved by:
- Ensuring consistent sparse format via .tocsr()
- Using user_row = X[user_idx, :] and user_row.indices for seen-item extraction

Once corrected, the model produced valid, stable recommendations.

---

## 📊 Final Results

After evaluation on held-out test interactions:

| Metric | Score |
|--------|-------|
| Recall@10 | 0.0119 |
| Recall@50 | 0.0370 |
| Novelty@10 | 0.9457 |
| Novelty@50 | 0.9564 |
| Category Diversity@10 | 0.8925 |
| Category Diversity@50 | 0.8070 |

---

## 🔍 Interpretation of Results

SLIM is optimized for sparser and more diverse recommendations rather than popularity-based recall.  
This explains the observed trade-offs:

- Low Recall  
Our dataset is based on AT_CartUpdate events, which are very sparse.  
Co-occurrence signals between items are weak → similarity learning is difficult.

- High Novelty & High Diversity  
Due to L1 regularization, SLIM learns sparse weights, which suppress global popularity and prioritize niche co-consumption patterns.  
This results in:
- Diverse recommendations
- Less popular, more surprising items being recommended

This behavior matches SLIM’s expected theoretical properties.

---

## 🎯 Conclusion

The SLIM implementation is correct and behaves as expected:
- It captures *meaningful but sparse* item similarity structure.
- It produces highly *novel and diverse* recommendations.
- Recall is low due to sparse interaction signal, not due to modeling error.

Further improvements (if desired) could involve:
- Using denser interaction types (Clicks / Views instead of CartUpdate)
- Tuning ElasticNet hyperparameters
- Blending SLIM recommendations with popularity-based reranking

### BPR with popularity sampling (1.5 points)

Now let's try to improve the matrix factorization models. In practice we tested alternating least squares and logistic factorization, which were trying to solve an optimization problem of predicting the correct score. But in recommender systems, we usually need to rank the items, and that is exactly what [BPR: Bayesian Personalized Ranking from Implicit Feedback](https://arxiv.org/pdf/1205.2618) does. Instead of predicting explicit ratings, BPR assumes that a user prefers items with which they have interacted over those they have not. It optimizes a pairwise ranking objective to ensure that, for each user, positive (observed) items are ranked higher than negative (unobserved) items, using gradient-based optimization (typically SGD, but there implementations that allow RMSProp, Adam, etc).

The BPR loss aims to maximize the probability that for a given user $u$, a positively interacted item $i$ is ranked higher than a negative item $j$. So, for a triplet $u, i, j$ the loss is defined as:

$$L(u, i, j) = -\ln \sigma(\hat{x}_{ui} - \hat{x}_{uj}) + \lambda ||\Theta||^2$$

where $\Theta$ represents parameters and adds for regularization. In our task, you'll be using a matrix factorization approach with user biases:

$$\hat{x}_{ui} = \langle p_u, q_u \rangle + b_{i}$$

##### Optimizer step

First, implement the optimizer step for a triplet $u, i, j$ - this will be your mini-batch in SGD optimization. For sampled indices, update the user embedding, item embedding and item bias. The derivation of the gradients for parameters is left to you (and is relatively easy to do).

Both in this and the next task reference solution contains some parts, that are JIT-compiled using `Numba` and seems to be working significantly faster. It is recommended to also adapt your code so that it's suitable for `Numba` (you'll have the templates for functions), but if you're lazy and patient - you can just send a pure Python version. But you still need to fully train the model at least once!

In [7]:
@nb.njit
def optimizer_step(user_vectors, item_vectors, user_idx, pos_item_idx, neg_item_idx, lr, lambd):
    """Optimization step for BPR model
    1. Calculate the dot-product scores for the positive and negative items.
    2. Apply sigmoid function to the difference of the scores to get P(u, i > j).
    3. Calculate the gradients for the user and item vectors.
    4. Update inplace the user and item vectors.

    Args:
        user_vectors (np.ndarray[n_users, dim]): User embeddings.
        item_vectors (np.ndarray[n_items, dim]): Item embeddings.
        user_idx (int): User index.
        pos_item_idx (int): Positive item index.
        neg_item_idx (int): Negative item index.
        lr (float): Learning rate.
        lambd (float): Regularization coefficient.
    """

    # first we're going to take references to curr param rows
    p_u = user_vectors[user_idx]
    q_i = item_vectors[pos_item_idx]
    q_j = item_vectors[neg_item_idx]

    # first, we're calculating the dot-product scores
    x_ui = np.dot(p_u, q_i)
    x_uj = np.dot(p_u, q_j)
    x_uij = x_ui - x_uj

    # take sigmoid of the difference from above
    sigma = 1.0 / (1.0 + np.exp(-x_uij))
    grad = 1.0 - sigma  # derivative of -log σ(x)

    # here caching old user vector before in-place update
    p_u_old = p_u.copy()

    p_u += lr * (grad * (q_i - q_j) - lambd * p_u)
    q_i += lr * (grad * p_u_old - lambd * q_i)
    q_j += lr * (-grad * p_u_old - lambd * q_j)

    # write back
    user_vectors[user_idx] = p_u
    item_vectors[pos_item_idx] = q_i
    item_vectors[neg_item_idx] = q_j

In [17]:
def test_bpr_optimizer(step_function):
    user_embeddings = np.array([
        [-0.004, -0.01 , -0.012,  0.038,  0.054],
        [ 0.012,  0.038,  0.   , -0.06 ,  0.015]
    ], dtype=np.float32)

    item_embeddings = np.array([
        [ 0.034,  0.047,  0.061, -0.043, -0.017],
        [-0.005,  0.065, -0.086,  0.055, -0.013]
    ], dtype=np.float32)

    gt_user_embeddings = np.array([
        [-0.00200446, -0.01080256, -0.00450913,  0.03270608,  0.05325943],
        [ 0.012     ,  0.038     ,  0.        , -0.06      ,  0.015     ],
    ], dtype=np.float32)

    gt_item_embeddings = np.array([
        [ 0.03345943,  0.04602858,  0.0597883,  -0.04066461, -0.01412233],
        [-0.00474943,  0.06485142, -0.0845383,   0.0525446,  -0.01557767],
    ], dtype=np.float32)

    optimizer_step(user_embeddings, item_embeddings, 0, 0, 1, 0.1, 0.1)

    assert np.allclose(user_embeddings, gt_user_embeddings), f"User embeddings are not close to the ground truth"
    assert np.allclose(item_embeddings, gt_item_embeddings), f"Item embeddings are not close to the ground truth"

    user_embeddings = np.array([
        [1, 0, 0, 0, 0],
        [1, 1, -1, -1, 0]
    ], dtype=np.float32)

    item_embeddings = np.array([
        [0, 1, 0, -1, 1],
        [-1, 0, 1, 0, -1]
    ], dtype=np.float32)

    gt_user_embeddings = np.array([
        [ 1.2689414 ,  0.26894143, -0.26894143, -0.26894143,  0.53788286],
        [ 1.        ,  1.        , -1.        , -1.        ,  0.        ]
    ], dtype=np.float32)

    gt_item_embeddings = np.array([
        [ 0.26894143,  1.        ,  0.        , -1.        ,  1.        ],
        [-1.2689414 ,  0.        ,  1.        ,  0.        , -1.        ]
    ], dtype=np.float32)

    step_function(user_embeddings, item_embeddings, 0, 0, 1, 1, 0)

    assert np.allclose(user_embeddings, gt_user_embeddings), f"User embeddings are not close to the ground truth"
    assert np.allclose(item_embeddings, gt_item_embeddings), f"Item embeddings are not close to the ground truth"

In [18]:
test_bpr_optimizer(optimizer_step)

##### Sampling
The next step is to implement the negative sampling. We will be storing IDs in the same format, as they are in dataset - just two arrays of IDs, where I-th interaction was between `user_ids[i]` and `item_ids[i]`. This will allow us to sample both users and items based on their "popularity" - the more times item or user was interacting with something, more frequently it will pop up in triplets. It's a good heuristic and works well in BPR.

In proposed implementation, we will be sampling any item as negative and then checking if it's actually negative with lookup into all user positives. Because there are a lot of items and dataset is sparse, the chance of sampling a fake negative is not really high, so we won't be slipping too much. Depending on implementation, storing all the positives, negatives and sampling from them may be slower or faster, but this one is definitely fast!

In [8]:
@nb.njit
def sample_triplet_popularity(item_ids: np.ndarray, user_ids: np.ndarray):
    """Sample triplet using fast popularity sampling from whole dataset, represented
    as two arrays of item and user IDs for each positive interaction from the dataset.
    1. Sample random interaction of the dataset
    2. Get user ID, positive item ID from this interaction.
    3. Sample random interaction of the dataset as negative item.
    The approach does not guarantee that the negative item is indeed negative for a user.

    Args:
        item_ids (np.ndarray[n_samples]): array of item IDs, each point represents a positive interaction from the dataset.
        user_ids (np.ndarray[n_samples]): array of user IDs, each point represents a positive interaction from the dataset.

    Returns:
        tuple[int, int, int]: user ID, positive item ID, negative item ID.
    """

    # first generate a random interaction of the dataset
    size = item_ids.shape[0]
    idx = np.random.randint(0, size)
    user_id = user_ids[idx]
    positive_id = item_ids[idx]

    # another item, not necessarily negative (de facto)
    negative_id = item_ids[np.random.randint(0, size)]

    return user_id, positive_id, negative_id

In [36]:
def test_sample_triplet_popularity(sampler, num_samples=100000):
    item_ids = np.array([4, 1, 2, 4, 3, 3, 4, 4, 2, 3])
    user_ids = np.array([1, 1, 2, 2, 3, 3, 4, 4, 3, 3])
    # ratings data would look something like this
    # pl.DataFrame({
    #     "user_id": [4, 1, 2, 4, 3, 3, 4, 4, 5, 5],
    #     "item_id": [1, 1, 2, 2, 3, 3, 4, 4, 5, 5],
    # })

    sampled_triplets = [sampler(item_ids, user_ids) for _ in range(num_samples)]
    # print('hello') for debugging purposes
    users, positives, negatives = zip(*sampled_triplets)

    gt_user_freqs = np.bincount(user_ids) / len(user_ids)
    gt_item_freqs = np.bincount(item_ids) / len(item_ids)
    user_freqs = np.bincount(users) / len(users)
    pos_item_freqs = np.bincount(positives) / len(positives)
    neg_item_freqs = np.bincount(negatives) / len(negatives)

    assert np.allclose(gt_user_freqs, user_freqs, atol=0.01)
    assert np.allclose(gt_item_freqs, pos_item_freqs, atol=0.01)
    assert np.allclose(gt_item_freqs, neg_item_freqs, atol=0.01)

In [37]:
test_sample_triplet_popularity(sample_triplet_popularity)

##### The actual model
Some code hints are left as comments, and docstring should provide the needed guidelines on how to implement the methods. Feel free to change the implementation if you want, as long as it's still BPR and it meets the quality benchmarks!

In [ ]:
from tqdm import trange
from collections import defaultdict


class PopularityBPR:
    def __init__(self,
                 dim: int = 128,
                 max_iter: int = 100,
                 learning_rate: float = 0.01,
                 lambd: float = 0.01,
                ):
        self.dim = dim
        self.max_iter = max_iter
        self.learning_rate = learning_rate
        self.lambd = lambd

    def _init_parameters(self, ratings: pl.DataFrame):
        """Initialize parameters for the model.
        1. Build mappings from user and item IDs to indices
        2. Initialize user and item vectors with scaled normal distribution
        3. Initialize item biases with zeros
        4. Set item biases to zero if there are no interactions for an item
        5. Set user vectors to zero if there are no interactions for a user

        Args:
            ratings (pl.DataFrame): pl.DataFrame with ratings
        """
        R, mappings = build_matrix_with_mappings(ratings, additive=False)
        self.user_id2idx, self.item_id2idx, self.user_idx2id, self.item_idx2id = mappings
        self.num_samples = len(ratings)
        self.n_users = len(self.user_id2idx)
        self.n_items = len(self.item_id2idx)
        self.item_biases = np.zeros((self.n_items, 1))
        self.user_vectors = np.random.normal(loc=0, scale=0.2 / self.dim, size=(self.n_users, self.dim))
        self.item_vectors = np.random.normal(loc=0, scale=0.2 / self.dim, size=(self.n_items, self.dim))
        user_has_no_interactions = R.sum(axis=1) == 0
        item_has_no_interactions = R.sum(axis=0) == 0
        self.item_vectors[item_has_no_interactions] = 0
        self.user_vectors[user_has_no_interactions] = 0


    def _init_sampler(self, ratings: pl.DataFrame) -> tuple[np.ndarray, np.ndarray]:
        """
        Convert ratings to the format, awaited by the sampling function - meaning
        np.arrays of item and user IDs for each positive interaction from the dataset.
        Also save the user positives in any lookup-friendly data structure for checking,
        whether and item is negative or positive for a user to skip it after sampling.

        Args:
            ratings (pl.DataFrame): Ratings data in standard format.
        Returns:
            tuple[np.ndarray, np.ndarray]: Item IDs and user IDs.
        """

        # TODO: your code here
        user_ids = ratings["user_id"].to_numpy() # because the sampler expects numpy
        item_ids = ratings["item_id"].to_numpy()

        self.user_pos = defaultdict(set)
        for user, item in zip(user_ids, item_ids):
          self.user_pos[user].add(item)

        return item_ids, user_ids


    def train_one_epoch(self):
        """Train one epoch of the model and compute the number of skipped triplets.
        For num_samples times:
        1. Sample a triplet (user, positive item, negative item)
        2. If the negative item is not a true negative for the user, skip the triplet
        3. Otherwise, update the model parameters

        Returns:
            int: Number of skipped triplets.
        """

        # TODO: your code here

        # skipped = 0
        # for _ in range(self.num_samples):
        #       triplet = sample_triplet()
        #       if verify_triplet(triplet):
        #           optimizer_step(triplet)
        # return skipped

        pass

    def fit(self, ratings: pl.DataFrame):
        """Fit the model to the ratings data.
        1. Initialize parameters and sampler
        2. Train the model for max_iter epochs
        3. Print the number of skipped triplets

        Args:
            ratings (pl.DataFrame): Ratings data in standard format.
        """

        # TODO: your code here

        pass

In [ ]:
bpr = PopularityBPR(max_iter=50)
bpr.fit(train)

For defined in this notebook train dataset, initialization and default hyperparameters from model class template, reference implementation achieves:

$$recall@10 >= 0.07$$
$$recall@50 >= 0.18$$

In [ ]:
user_embeddings, item_embeddings = extract_embeddings(bpr)
knn = DotProductKNN(user_embeddings, item_embeddings)

# TODO: your code here
# compare speed and recall, better to write your thoughts as md cell

### BPR with adaptive sampling (1.5 points)

Now let's spice things up a bit!

Popularity sampling is fast and effective, but eventually it converges to a state where it primarily samples easy, obvious negatives and does not learn anything new. This can be fixed by using hard negatives - the ones, that are negative for user, but have high model score. We'll do this by changing popularity sampling to weighted sampling, where weights are proportionate to model confidence.

For positive weights, we'll compute the probabilites as frequencies in dataset for each user (or any other way if you want to). For first `num_epochs_without_updates` we'll set the negative weights uniform, after that we'll compute the scores and softmax them with temperature to get the sampling probabilities. The higher the score - the more frequently we will be sampling these negatives.

Since we're using complex weighted sampling (three random generator calls in total), we'll ensure that both positive and negative weights are correct, meaning that negative items have zero weight in positive sampling weights and vice versa. This should be done both in sampler initialization and during weights update.

In [ ]:
@nb.njit
def nb_random_choice(arr, prob):
    """Analogue of weighted np.random.choice (Numba-compatible).
    Args:
        arr (np.ndarray): Array of items to sample from.
        prob (np.ndarray): Array of probabilities for each item.
    Returns:
        int: Index of the sampled item.
    """
    return arr[np.searchsorted(np.cumsum(prob), np.random.random(), side="right")]


@nb.njit
def weighted_sample_triplet(users, items, user_weights, positive_weights, negative_weights):
    """Weighted sampling of triplet (user, positive item, negative item).

    Args:
        users (np.ndarray): Array of user IDs.
        items (np.ndarray): Array of item IDs.
        user_weights (np.ndarray[n_users]): Array of user weights.
        positive_weights (np.ndarray[n_users, n_items]): Array of positive item weights for each user.
        negative_weights (np.ndarray[n_users, n_items]): Array of negative item weights for each user.

    Returns:
        tuple[int, int, int]: User ID, positive item ID, negative item ID.
    """
    user_idx = nb_random_choice(users, user_weights)
    positive_item_idx = nb_random_choice(items, positive_weights[user_idx])
    negative_item_idx = nb_random_choice(items, negative_weights[user_idx])
    return user_idx, positive_item_idx, negative_item_idx

In [ ]:
class AdaptiveBPR(PopularityBPR):
    def __init__(self,
                 dim: int = 128,
                 max_iter: int = 100,
                 learning_rate: float = 0.01,
                 lambd: float = 0.01,
                 weight_temperature: float = 0.5,
                 num_epochs_without_updates: int = 10
                ):
        super().__init__(dim, max_iter, learning_rate, lambd)
        self.t = weight_temperature
        self.num_epochs_without_updates = num_epochs_without_updates

    def _init_sampler(self, ratings: pl.DataFrame):
        """Initialize sampler weights.
        There should be three weights, all proportional to popularity in dataset:
        1. User weights - np.array[n_users]
        2. Positive item weights - np.array[n_users, n_items]
        3. Negative item weights - np.array[n_users, n_items]
        Do not forget to normalize item weights after initialization inside each row and all user weights.

        Args:
            ratings (pl.DataFrame): Ratings data in standard format.
        """

        # TODO: your code here

        # user_weights = [0, 0, 0, ...]
        # positive_weights = [[0, 0, 0, ...], [0, 0, 0, ...], ...]

        # for user, item in dataset:
        #     user_weights[user] += 1
        #     positive_weights[user, item] += 1

        pass

    def train_one_epoch(self):
        """
        For num_samples times:
        1. Sample a triplet (user, positive item, negative item) using weighted negative sampling
        3. Update the model parameters

        Returns:
            int: Number of skipped triplets.
        """

        # TODO: your code here

        pass

    def update_weights(self):
        """Update negative item weights based on model scores.
        1. Compute the scores for all items for each user.
        2. Set to zero the weights for positive items, that are not negative for the user.
        3. Apply softmax with temperature to the negative weights.
        """

        # TODO: your code here

        # scores = user_emb @ item_emb
        # negative_mask = self.negative_weights > 0
        # negative_weights = softmax(scores[negative_mask], t)

        pass

    def fit(self, ratings: pl.DataFrame):
        """Fit the model to the ratings data.
        - First initialize parameters
        - Initialize sampler and weights
        - For max_iter epochs:
            - Train the model for one epoch
            - If epoch >= num_epochs_without_updates, update the weights

        Args:
            ratings (pl.DataFrame): Ratings data in standard format.
        """

        # TODO: your code here

        pass

In [ ]:
bpr = AdaptiveBPR()
bpr.fit(train)

For defined in this notebook train dataset, initialization and default hyperparameters from model class template, reference implementation achieves:

$$recall@10 >= 0.09$$
$$recall@50 >= 0.20$$

In [ ]:
user_embeddings, item_embeddings = extract_embeddings(bpr)
knn = DotProductKNN(user_embeddings, item_embeddings)

# TODO: your code here
# compare speed and recall, better to write your thoughts as md cell

### Quality benchmark (0.5 points)

This part is supposed to inspire you to be creative!

Come up with a benchmark setup, where you compare:
- The speed of model training on different dataset sizes (at least three).
- The memory of stored CGs (ANN index, `.npz` file or sparse/dense I2I matrix).
- The speed of CGs on inference (request per second in non-batched format).
- The $recall@10$ and $recall@50$ of the algorithms you got.

Each metric gets you 0.1 point, except model training speed which gets you 0.2 points. Present your results in a human-readable format, preferably using pandas/polars DataFrames or various plots within the notebook. Reports in an unreadable format may receive fewer points.

In [ ]:
# TODO: your code here